### PR TITLE
Fix #2276

### LINKED ISSUE #2276: Trying to download a 'html summary dir' filepath type will fail
Because the download handler doesn't support directories ATM. It's possible to modify as it is done for the `DownloadStudyBIOMSHandler`

### PR DESCRIPTION
**Pull request by josenavas**
Fixes the download handler to support directories.

### CODE REVIEW

File: qiita_pet/handlers/download.py [hunk 1]
```diff
@@ -37,19 +38,42 @@ def get(self, filepath_id):
...
+
+            all_files = '\n'.join(
+                ["- %s /protected/%s %s" % (getsize(fp), sfp, n)
+                 for fp, sfp, n in to_download])
+
+            self.set_header('X-Archive-Files', 'zip')
+            self.write("%s\n" % all_files)
+            fname = '%s.zip' % fname
+        else:
+            # If we don't have nginx, write a file that indicates this
```
Comment (antgonza) [inline qiita_pet/handlers/download.py]: A lot of this code is copy pastes from other classes, would you mind creating a new class from which these functions are inherit? Perhaps use the BaseHandlerDownload handler ...

**Comment by josenavas** [comment]
@antgonza I factored out as much as code as I found possible.

**Comment by antgonza** [comment]
thanks! Looks good to me except that `self.write("This installation of Qiita was not equipped with " ...` can be made into a single method, right? Currently is used in 2 different classes.

File: qiita_db/util.py [hunk 2]
```diff
@@ -845,6 +845,44 @@ def move_filepaths_to_upload_folder(study_id, filepaths):
...
+                        artifact_id
+                 FROM qiita.filepath
+                    JOIN qiita.filepath_type USING (filepath_type_id)
+                    JOIN qiita.data_directory USING (data_directory_id)
+                    LEFT JOIN qiita.artifact_filepath USING (filepath_id)
+                 WHERE filepath_id = %s"""
+        qdb.sql_connection.TRN.add(sql, [filepath_id])
+        res = dict(qdb.sql_connection.TRN.execute_fetchindex()[0])
+
+        def path_builder(db_dir, filepath, mountpoint, subdirectory, obj_id):
```
Comment (wasade) [inline qiita_db/util.py]: Is a nested function necessary? It looks like it only adds complexity here...?

File: qiita_pet/handlers/download.py [hunk 3]
```diff
@@ -77,6 +46,149 @@ def _generate_files(self, header_name, accessions, filename):
...
+            Path to the directory
+
+        Returns
+        -------
+        list of (str, str, str)
+            The path information needed by nginx for each file in the
+            directory
+        """
+        basedir = get_db_files_base_dir()
+        basedir_len = len(basedir) + 1
```
Comment (wasade) [inline qiita_pet/handlers/download.py]: Is this, and its use below, basically the same as `os.dirname`? Same comment for line 89 and its use in the next function

File: qiita_pet/handlers/download.py [hunk 4]
```diff
@@ -77,6 +46,149 @@ def _generate_files(self, header_name, accessions, filename):
...
+        to_download = []
+        for i, (fid, path, data_type) in enumerate(artifact.filepaths):
+            # ignore if tgz as they could create problems and the
+            # raw data is in the folder
+            if data_type == 'tgz':
+                continue
+            if data_type == 'directory':
```
Comment (wasade) [inline qiita_pet/handlers/download.py]: Would it be better to ask the filesystem if the path is a directory?

File: qiita_db/util.py (hunk 2 above)
Comment (josenavas) [reply qiita_db/util.py]: I moved it outside the function cause it is used in another function, to reduce code duplication.

File: qiita_pet/handlers/download.py (hunk 3 above)
Comment (josenavas) [reply qiita_pet/handlers/download.py]: No, this is used to remove a prefix from the path that can be formed by multiple dirs.

File: qiita_pet/handlers/download.py (hunk 4 above)
Comment (josenavas) [reply qiita_pet/handlers/download.py]: done

**Pull request merged**